Move TranslateExpression above SimplifyRowExpression

hellium01 opened the pull request:
On top of #13047 (after commit "Move TranslateExpressions above last projectionPushDown"). 
```
== NO RELEASE NOTE ==

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java:
```diff
@@ -242,6 +243,9 @@ public Expression toExpression(Object object, Type type)
 
     public static boolean isSupportedLiteralType(Type type)
     {
+        if (type instanceof FunctionType) {
```
Add a comment to this

hellium01 replied:
Another way is to catch `UnsupportedOperationException` from `getJavaType` but I prefer to list it out explicitly instead.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1246,6 +1246,9 @@ private PhysicalOperation visitScanFilterAndProject(
                 }
             }
             catch (RuntimeException e) {
```
nit
```java
catch (PrestoException e) {
    throw e;
}
catch (RuntimeException e) {
    throw new PrestoException(COMPILER_ERROR, "Compiler failed", e);
}

highker commented on presto-main/src/main/java/com/facebook/presto/sql/relational/optimizer/ExpressionOptimizer.java:
```diff
@@ -56,6 +56,7 @@
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
+@Deprecated
 public class ExpressionOptimizer
```
- This class is not used by any prod code anymore. Consider moving it as a test class `TestingExpressionOptimizer`.
- Actually rather than making it in test scope, we should *completely* remove it. `ExpressionOptimizer` is used by some critical tests like `TestExpressionInterpreter`. It is dangerous to have this class not running in production yet checking some critical paths. Behavior will become inconsistent if we still rely on this for tests.

hellium01 replied:
Yes, removing it will be the goal, but I haven't included it in this PR.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -0,0 +1,92 @@
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType())) {
+                return logicalRowExpressions.pushNegationToLeaves(node);
```
`pushDownNegations` is more sophisticated than `pushNegationToLeaves`. Could you check the consistency between these two and add tests?

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTranslateExpressions.java:
```diff
@@ -62,23 +61,27 @@
-                Optional.empty());
-        Aggregation translated = translator.translateAggregation(aggregation, TEST_SESSION, TypeProvider.viewOf(ImmutableMap.of("input", INTEGER)));
-        assertEquals(translated, new Aggregation(
+        PlanNode result = tester().assertThat(new TranslateExpressions(METADATA, new SqlParser()).aggreagationRowExpressionRewriteRule())
```
typo: `aggreagationRowExpressionRewriteRule`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TranslateExpressions.java:
```diff
@@ -0,0 +1,187 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Collections.emptyList;
+
+public class TranslateExpressions
```
what's the intuition behind this inheritance? `TranslateExpressions` should be a clean and standalone module. Other optimizer rules will not call it.

hellium01 replied:
`TranslateExpressions` is a special case of  `RowEpxressionRewriteRuleSet` that rewrites `OriginalExpression` to `RowExpression`. This ensures us to share the same common code and prevents situations that we had that some planNode was not added in some places for `Expressions`.

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralInterpreter.java:
```diff
@@ -158,6 +159,9 @@ public static Object evaluate(ConnectorSession session, ConstantExpression node)
         if (type instanceof IntervalYearMonthType) {
             return new SqlIntervalYearMonth(((Long) node.getValue()).intValue());
         }
+        if (type.getJavaType().equals(Slice.class)) {
```
Without this, we will print indeterministic string for these field in the plan, which cause PlanDeterministic tests fail.

highker commented:
@hellium01, could you rebase?

hellium01 commented:
Thanks, rebased on the current head.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -0,0 +1,92 @@
+import static com.facebook.presto.spi.relation.ExpressionOptimizer.Level.SERIALIZABLE;
+import static java.util.Objects.requireNonNull;
+
+public class SimplifyRowExpressions
```
Could you make sure
- `TestSimplifyExpressions` covers your new code and
- it has the exact same behavior as `SimplifyExpressions`?

hellium01 commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -241,20 +248,20 @@ public RowExpression combineDisjunctsWithDefault(Collection<RowExpression> expre
+     * <p>
      * An applicable example:
-     *
-     *        NOT
```
It is in airlift's IntelliJ style file but will disable adding <p> for empty lines.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -291,9 +292,9 @@ public static PlanNode pushPredicateIntoTableScan(
-                            Map.Entry::getKey,
-                            entry -> new VariableReferenceExpression(getColumnName(session, metadata, node.getTable(), entry.getValue()), entry.getKey().getType())));
-            RowExpression translatedPredicate = replaceExpression(SqlToRowExpressionTranslator.translate(predicate, predicateTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), session, false), symbolToColumnMapping);
+                            entry -> new VariableReferenceExpression(entry.getKey().getName(), types.get(new SymbolReference(entry.getKey().getName()))),
```
Why these two type changes?

highker commented on presto-main/src/test/java/com/facebook/presto/sql/TestRowExpressionSerde.java:
```diff
@@ -164,15 +163,14 @@ public void testArrayGet()
-                                        DOUBLE)),
-                        constant(1L, INTEGER),
-                        constant(1.1, DOUBLE)));
+        RowExpression rowExpression = getRoundTrip("ROW(1, 1.1)", true);
```
Why do we need to rewrite the logic?

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -273,15 +280,15 @@ public RowExpression pushNegationToLeaves(RowExpression expression)
-     *    OR   c   NOT              a    b   NOT  d
-     *   /  \        |                         |
-     *  a    b       d                         c
+     * NOT
```
You really need to fix your IntelliJ.....

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -294,15 +301,15 @@ public RowExpression convertToConjunctiveNormalForm(RowExpression expression)
-     *    OR   c   NOT                       NOT  d         NOT  d
-     *   /  \        |                         |              |
-     *  a    b       d                         c              c
+     * NOT                                    OR
```
same here

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -442,6 +470,19 @@ private boolean isNegationExpression(RowExpression expression)
+        private OperatorType getOperator(RowExpression expression)
+        {
+            if (expression instanceof CallExpression) {
+                return functionResolution.getOperator(((CallExpression) expression).getFunctionHandle()).get();
```
`.get()` will error out if the call expression does not have an operator

hellium01 commented:
Squashed the last two commit but git made it even harder to look, let me know if you want me to split them :).

hellium01 commented:
explain verifier tests also passed with 100k queries.

highker commented:
@hellium01, please request my review instead of commenting the PR. Otherwise, I will miss it.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,278 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+     * Cartesian cross product of List of List.
+     * For example, [[a], [b, c], [d]] becomes [[a,b,d], [a,c,d]]
+     */
+    private List<List<RowExpression>> crossProduct(List<List<RowExpression>> lists)
```
static

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,278 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+        return result;
+    }
+
+    private List<List<RowExpression>> crossProduct(List<List<RowExpression>> left, List<RowExpression> right)
```
static

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,278 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+                currentForm = flip(currentForm);
+            }
+
+            int currentSize = subPredicates.stream()
```
nit: oneline

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,278 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            int expectedSize;
+            try {
+                expectedSize = multiplyExact(
+                        subPredicates.stream()
```
nit: oneline

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -327,6 +327,13 @@ public RowExpression convertToDisjunctiveNormalForm(RowExpression expression)
         return convertToNormalForm(expression, OR);
     }
 
+    public RowExpression convertToSmallestNormalForm(RowExpression expression)
```
`MinimalNormalForm`

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -625,6 +632,16 @@ private boolean isComparisonExpression(RowExpression expression)
+    private int numOfSubPredicates(RowExpression expression)
+    {
+        if (expression instanceof SpecialFormExpression) {
+            return getSubPredicates((SpecialFormExpression) expression).stream()
```
nit: oneline

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestSimplifyExpressions.java:
```diff
@@ -114,13 +130,24 @@ public void testExtractCommonPredicates()
                 normalize(expectedExpression));
+        TestingRowExpressionTranslator translator = new TestingRowExpressionTranslator(METADATA);
+        RowExpression actualRowExpression = translator.translate(actualExpression, TypeProvider.viewOf(TYPES));
+        RowExpression simplifiedByRowExpression = SimplifyRowExpressions.rewrite(actualRowExpression, METADATA, TEST_SESSION.toConnectorSession());
```
I don't think `By` actually mean something.... Remove it from the variable name. Same for all other `By`s.

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestSimplifyExpressions.java:
```diff
@@ -37,14 +47,20 @@
+import static java.util.stream.Collectors.toMap;
 import static org.testng.Assert.assertEquals;
 
 public class TestSimplifyExpressions
```
add the following two tests to `testExtractCommonPredicates`
```java
assertSimplifies("true", "true");
assertSimplifies("if (X, X, Y)", "if (X, X, Y)");

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -0,0 +1,132 @@
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Boolean isRoot, RowExpressionTreeRewriter<Boolean> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType()) && functionResolution.isNotFunction(node.getFunctionHandle())) {
```
`BooleanType.BOOLEAN.equals(node.getType())` should be an assertion instead of a condition. Put a checkState

hellium01 replied:
Probably we should just remove it? We should not ever have none boolean type for NOT/AND/OR?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -0,0 +1,132 @@
+        @Override
+        public RowExpression rewriteSpecialForm(SpecialFormExpression node, Boolean isRoot, RowExpressionTreeRewriter<Boolean> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType()) && isConjunctiveDisjunctive(node.getForm())) {
```
`BooleanType.BOOLEAN.equals(node.getType())` should be an assertion instead of a condition. Put a checkState

wenleix commented on presto-main/src/test/java/com/facebook/presto/sql/relational/TestLogicalRowExpressions.java:
```diff
@@ -50,6 +55,14 @@
 {
     private FunctionManager functionManager;
     private LogicalRowExpressions logicalRowExpressions;
+    private final RowExpression a = name("a");
```
nit: static

wenleix commented on presto-main/src/test/java/com/facebook/presto/sql/relational/TestLogicalRowExpressions.java:
```diff
@@ -158,6 +152,92 @@ public void testPushNegationToLeaves()
+    public void testEliminateConstant()
+    {
+        // Testing eliminate constant
+        assertEquals(logicalRowExpressions.convertToConjunctiveNormalForm(or(and(TRUE_CONSTANT, a), and(FALSE_CONSTANT, b))),
```
nit: I would put one new line for each parameter, i.e.

```java
        assertEquals(
               logicalRowExpressions.convertToConjunctiveNormalForm(or(and(TRUE_CONSTANT, a), and(FALSE_CONSTANT, b))),
                a);
```

ditto for other tests

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,278 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+    private static class ConvertNormalFormVisitorContext
+    {
+        private final Form clauseJoiner;
+        private final int levelFromRoot;
```
nit: What about `depth` ?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -0,0 +1,132 @@
+
+        private RowExpression rewrite(RowExpression expression, ConnectorSession session)
+        {
+            RowExpression optimized = optimizer.optimize(expression, SERIALIZABLE, session);
```
nit: `optimizedRowExpression`. We usually use `optimized` for boolean variable.

highker commented:
@hellium01, I don't extra comment on this. Maybe chat with @wenleix to get his approval.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            RowExpression rewritten = combinePredicates(
+                    specialFormExpression.getForm(),
+                    extractPredicates(specialFormExpression.getForm(), specialFormExpression)
+                            .stream()
```
nit: merge with previous line

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            return predicates;
+        }
+        // initialize to self
+        int[] reduceTo = IntStream.range(0, predicates.size()).toArray();
```
nit: never a big fan of `IntStream` and `toArray()`

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+                        reduceTo[i] = j; //prefer smaller set
+                    }
+                    else if (isSameSet(predicates.get(reduceTo[i]), predicates.get(j))) {
+                        reduceTo[i] = Integer.min(reduceTo[i], j); //prefer predicates that appears earlier.
```
nit: i think we usually use `Math.min` (and statically import)

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
-            // normalize arguments
-            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormalFormVisitor(), clauseJoiner);
-            RowExpression right = specialForm.getArguments().get(1).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+            //extractCommonPredicates when currentForm is same as clauseJoiner can produce opposite form
```
nit: space

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            // The end result can be a cartesian cross product list of list
+            int expectedSize;
+            try {
+                expectedSize = multiplyExact(subPredicates.stream().mapToInt(List::size).reduce(1, Math::multiplyExact), subPredicates.size());
```
As discussed offline, use  for-loop so we can return early when `expectedSize` is too large. 

Using exception to handle normal business logic is not a good practice and should be avoided.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            return null;
+        }
+        Set<RowExpression> commonPredicates = new LinkedHashSet<>(predicates.get(0));
+        predicates.forEach(commonPredicates::retainAll);
```
nit: use a for loop starting from 1

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+                .stream()
+                .map(subPredicates ->
+                        combinePredicates(flip(rootForm),
+                                subPredicates
```
Looks like here (line 677 - 679) is implementing `removeAll` with stream. Maybe can replace with a simple function call?

hellium01 replied:
`removeAll` is in-place modification, which we probably don't want here.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            return null;
+        }
+        List<RowExpression> remainingPredicates = unmodifiableList(predicates
+                .stream()
```
nit: merge to previous line.

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+        List<List<RowExpression>> result = new ArrayList<>();
+        for (List<RowExpression> combined : left) {
+            for (RowExpression toAdd : right) {
+                List<RowExpression> newLine = new ArrayList<>(combined);
```
nit: `newClause`? 

Also maybe name `combined`  as `previousClause`, and `toAdd` as `newTerm`. So we combine `previousClause` and `newTerm` into `newClause`

wenleix commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java:
```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            case OR:
+                return AND;
+        }
+        throw new UnsupportedOperationException("Invalid binary logical operation");
```
nit: We usually do
```java
throw new UnsupportedOperationException("Invalid binary logical operation: " + binaryLogicalOperation);
```

wenleix commented:
This PR is required to fix https://github.com/prestodb/presto/issues/13360

hellium01 commented:
Thanks for reviewing. I addressed the comments and updated the code in commit "Improve ConvertNormalForm" to make the code a little bit easier to understand.
